app/models/leap_year.rb

```diff
@@ -0,0 +1,15 @@
+class LeapYear
+  def initialize(date)
+    @date = date
+  end
+
+  def leap_year?
+    year = @date.year
+
+    divisible_by_4 = year % 4 == 0
+    not_divisible_by_100 = year % 100 != 0
+    divisible_by_400 = year % 400 == 0
+
+    (divisible_by_4 && not_divisible_by_100) || divisible_by_400
+  end
+end
```

Comment:
On the topic of design - we always need to be iterating on our solutions.

At first, we wanted to solve the problem of this logic being defined in multiple places. This class does solve this, and well.

The next step is thinking about whether we can tighten the interface for this class up.

Currently, the class receives a date (EG: 19th May 2021) to determine whether the year of that date is a leap year.

I think the next thing to think about is: Will this class be even more useful if we pass the year itself instead of the full date?

The upside is: We have more flexibility in what can be passed through. You can see in your tests for this class that you need to create a date to pass in. With the updated interface  you could pass an integer that looks like a year through (EG: 2021).

The downside is: Callers of this method (EG: Book and Author) will need to pull a year out of a date object, instead of just doing it in this class.

At all times during development you'll have to make 100s of small decisions like these, and their impacts add up.

In this case, I think we're fine to pass the full date in. If we later find a use case where we only have a year, we'll need to adjust the interface for this class.

The only reason I'm writing this is to give you an idea about the kinds of things you might want to consider.